subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/msvcpp/DefaultUcrtLocator.java

```diff
@@ -16,234 +16,45 @@
-
-        List<String> result = new ArrayList<String>();
-        for (String version : ucrtVersions) {
-            File inc = includeDirs.get(version);
-            File lib = includeDirs.get(version);
-            if (inc != null && lib != null) {
-                File ucrtInc = new File(inc, "ucrt");
-                File ucrtLib = new File(lib, "ucrt");
-                if (ucrtInc.isDirectory() && ucrtLib.isDirectory()) {
-                    result.add(version);
-                }
-            }
-        }
-        return result.toArray(new String[result.size()]);
+    public String getComponentName() {
+        return COMPONENT_NAME;
     }
 
-    private void putUcrt(Ucrt ucrt) {
-        Set<Ucrt> ucrts = foundUcrts.get(ucrt.getBaseDir());
-        if (ucrts == null) {
-            ucrts = new HashSet<Ucrt>();
-            foundUcrts.put(ucrt.getBaseDir(), ucrts);
-        }
-        ucrts.add(ucrt);
+    @Override
+    String getDisplayName() {
+        return DISPLAY_NAME;
     }
 
-    private Ucrt getBestUcrt(File baseDir) {
-        final SortedSet<Ucrt> candidates = new TreeSet<Ucrt>(new DescendingUcrtVersionComparator());
-        Set<Ucrt> ucrts = foundUcrts.get(baseDir);
-        if (ucrts != null) {
-            candidates.addAll(ucrts);
-        }
-        return candidates.isEmpty() ? null : candidates.iterator().next();
+    @Override
+    boolean isValidComponentBaseDir(File baseDir) {
+        // Nothing special to check for UCRT
+        return true;
     }
 
-    private Ucrt getBestUcrt() {
-        final SortedSet<Ucrt> candidates = new TreeSet<Ucrt>(new DescendingUcrtVersionComparator());
-        for (Set<Ucrt> ucrts : foundUcrts.values()) {
-            candidates.addAll(ucrts);
-        }
-        return candidates.isEmpty() ? null : candidates.iterator().next();
-    }
-
-    private static class UcrtFound implements SearchResult {
-        private final Ucrt ucrt;
-
-        public UcrtFound(Ucrt ucrt) {
-            this.ucrt = ucrt;
-        }
-
-        public Ucrt getUcrt() {
-            return ucrt;
-        }
-
-        public boolean isAvailable() {
-            return true;
-        }
-
-        public void explain(TreeVisitor<? super String> visitor) {
-        }
+    @Override
+    boolean isValidComponentIncludeDir(File includeDir) {
+        return new File(includeDir, "io.h").exists();
     }
 
-    private static class UcrtNotFound implements SearchResult {
-        private final String message;
-
-        private UcrtNotFound(String message) {
-            this.message = message;
-        }
-
-        public Ucrt getUcrt() {
-            return null;
-        }
-
-        public boolean isAvailable() {
-            return false;
-        }
-
-        public void explain(TreeVisitor<? super String> visitor) {
-            visitor.node(message);
-        }
+    @Override
+    boolean isValidComponentLibDir(File libDir) {
+        return new File(libDir, "x86/libucrt.lib").exists();
```

Comment:
Do we know if it's safe to assume that all targeting architectures are present in the installation when we find x86 folder? This would need to be validated or else we could have a UCRT that only has x86 libraries and we are targeting an x64 platform.

Reply:
That's a good question.  I wanted to do more validation than just "does this directory exist?"  And it doesn't seem realistic to validate everything.  So it's a question of how much is enough?  I was just using the bar that was set before (that at least one platform library exists) but I'm not opposed to say checking for the libraries for all platforms that we support and rejecting an installation that has say, and x86 library but not a x64 library.  WDYT?

Reply:
I went ahead and added validation for both x86 and x64 libraries.